internal/extender/resourcereservations.go

```diff
@@ -15,26 +15,281 @@
+		return rrm.bindExecutorToSoftReservation(ctx, executor, node)
+	}
+
+	return werror.ErrorWithContextParams(ctx, "failed to find free reservation for executor")
+}
+
+// GetReservedResources returns the resources per node that are reserved for executors.
+func (rrm *ResourceReservationManager) GetReservedResources() resources.NodeGroupResources {
+	resourceReservations := rrm.resourceReservations.List()
+	usage := resources.UsageForNodes(resourceReservations)
+	usage.Add(rrm.softReservationStore.UsedSoftReservationResources())
+	return usage
+}
+
+func (rrm *ResourceReservationManager) bindExecutorToResourceReservation(ctx context.Context, executor *v1.Pod, reservationName string, node string) error {
+	resourceReservation, ok := rrm.GetResourceReservation(executor)
+	if !ok {
+		return werror.ErrorWithContextParams(ctx, "failed to get resource reservationName")
+	}
+	copyResourceReservation := resourceReservation.DeepCopy()
+	reservationObject := copyResourceReservation.Spec.Reservations[reservationName]
+	reservationObject.Node = node
+	copyResourceReservation.Spec.Reservations[reservationName] = reservationObject
+	copyResourceReservation.Status.Pods[reservationName] = executor.Name
+	err := rrm.resourceReservations.Update(copyResourceReservation)
+	if err != nil {
+		return werror.WrapWithContextParams(ctx, err, "failed to update resource reservationName")
+	}
+	return nil
+}
+
+func (rrm *ResourceReservationManager) bindExecutorToSoftReservation(ctx context.Context, executor *v1.Pod, node string) error {
+	driver, err := rrm.podLister.getDriverPod(ctx, executor)
+	if err != nil {
+		return err
+	}
+	sparkResources, err := sparkResources(ctx, driver)
+	if err != nil {
+		return err
+	}
+	softReservation := v1beta1.Reservation{
+		Node:   node,
+		CPU:    sparkResources.executorResources.CPU,
+		Memory: sparkResources.executorResources.Memory,
+	}
+	return rrm.softReservationStore.AddReservationForPod(ctx, driver.Labels[common.SparkAppIDLabel], executor.Name, softReservation)
+}
+
+func (rrm *ResourceReservationManager) getUnboundReservations(ctx context.Context, executor *v1.Pod) (map[string]string, error) {
+	resourceReservation, ok := rrm.GetResourceReservation(executor)
+	if !ok {
+		return nil, werror.ErrorWithContextParams(ctx, "failed to get resource reservation")
+	}
+	activePodNames, err := rrm.getActivePodNames(ctx, executor)
+	if err != nil {
+		return nil, err
+	}
+
+	unboundReservationsToNodes := make(map[string]string, len(resourceReservation.Spec.Reservations))
+	for reservationName, reservation := range resourceReservation.Spec.Reservations {
+		podIdentifier, ok := resourceReservation.Status.Pods[reservationName]
+		if !ok || !activePodNames[podIdentifier] {
+			unboundReservationsToNodes[reservationName] = reservation.Node
+		}
+	}
+	return unboundReservationsToNodes, nil
+}
+
+func (rrm *ResourceReservationManager) getFreeExtraExecutorSpots(ctx context.Context, executor *v1.Pod) (int, error) {
```

Comment:
I think instead of introducing the concept of `ExtraExecutors` you should use the existing terminology, either `getSoftReservationExecutorSpots` or `getDynamicAllocatedExecutorSpots`